File: src/test/java/io/pivotal/kotlindemo/java/accounts/AccountRepositoryTest.java
package io.pivotal.kotlindemo.java.accounts;

import io.pivotal.kotlindemo.java.clients.Client;
import org.junit.Before;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public abstract class AccountRepositoryTest {
    public abstract AccountRepository getAccountRepository();

    private AccountRepository accountRepository;

    private static final Client client1 = new Client(
            123L,
            "client1first",
            "client1last",
            "client1@email.com"
    );

    private static final Client client2 = new Client(
            456L,
            "client2first",
            "client2last",
            "client2@email.com"
    );

    @Before
    public void setup() {
        accountRepository = getAccountRepository();
    }

    @Test
    public void create_addsAccountsWithUniqueIds() {
        CreateAccountRequest createAccountRequest1 = new CreateAccountRequest(
                client1.getId(),
                "account1 number",
                "account1 full name",
                "account1 nickname",
                AccountType.CHECKING);

        Account account1 = accountRepository.create(createAccountRequest1);

        assertThat(account1.getId()).isNotNull();
        assertThat(account1).isEqualToIgnoringGivenFields(
                new Account(
                        null,
                        createAccountRequest1.getClientId(),
                        createAccountRequest1.getNumber(),
                        createAccountRequest1.getFullname(),
                        createAccountRequest1.getNickname(),
                        createAccountRequest1.getType()
                ),
                "id"
        );

        CreateAccountRequest createAccountRequest2 = new CreateAccountRequest(
                client2.getId(),
                "account2 number",
                "account2 full name",
                "account2 nickname",
                AccountType.CHECKING);

        Account account2 = accountRepository.create(createAccountRequest2);

        assertThat(account2.getId()).isNotNull();
        assertThat(account2).isEqualToIgnoringGivenFields(
                new Account(
                        null,
                        createAccountRequest2.getClientId(),
                        createAccountRequest2.getNumber(),
                        createAccountRequest2.getFullname(),
                        createAccountRequest2.getNickname(),
                        createAccountRequest2.getType()
                ),
                "id"
        );

        assertThat(account1.getId()).isNotEqualTo(account2.getId());
    }

    @Test
    public void findAll_returnsAllCreatedAccounts() {
        CreateAccountRequest createAccountRequest1 = new CreateAccountRequest(
                client1.getId(),
                "account1 number",
                "account1 full name",
                "account1 nickname",
                AccountType.CHECKING);

        Account account1 = accountRepository.create(createAccountRequest1);

        CreateAccountRequest createAccountRequest2 = new CreateAccountRequest(
                client2.getId(),
                "account2 number",
                "account2 full name",
                "account2 nickname",
                AccountType.CHECKING);

        Account account2 = accountRepository.create(createAccountRequest2);

        assertThat(accountRepository.findAll()).containsExactlyInAnyOrder(
                account1,
                account2
        );
    }

    @Test
    public void findById_whenTheAccountExists_returnsTheAccountWithTheGivenId() {
        CreateAccountRequest createAccountRequest1 = new CreateAccountRequest(
                client1.getId(),
                "account1 number",
                "account1 full name",
                "account1 nickname",
                AccountType.CHECKING);

        Account account1 = accountRepository.create(createAccountRequest1);

        CreateAccountRequest createAccountRequest2 = new CreateAccountRequest(
                client2.getId(),
                "account2 number",
                "account2 full name",
                "account2 nickname",
                AccountType.CHECKING);

        Account account2 = accountRepository.create(createAccountRequest2);

        assertThat(accountRepository.findById(account1.getId())).isEqualTo(account1);
        assertThat(accountRepository.findById(account2.getId())).isEqualTo(account2);
    }

    @Test(expected = NoAccountFoundException.class)
    public void findById_whenTheAccountDoesNotExist_throwsNoAccountFoundException() {
        Long nonExistantAccountId = 12345L;
        accountRepository.findById(nonExistantAccountId);
    }

    @Test
    public void findByClientId_returnsAPossiblyEmptyListOfAllClientAccountsForTheGivenClientId() {
        CreateAccountRequest createAccountRequest1 = new CreateAccountRequest(
                client1.getId(),
                "account1 number",
                "account1 full name",
                "account1 nickname",
                AccountType.CHECKING);

        Account account1 = accountRepository.create(createAccountRequest1);

        CreateAccountRequest createAccountRequest2 = new CreateAccountRequest(
                client1.getId(),
                "account2 number",
                "account2 full name",
                "account2 nickname",
                AccountType.CHECKING);

        Account account2 = accountRepository.create(createAccountRequest2);

        assertThat(accountRepository.findByClientId(client1.getId())).containsExactlyInAnyOrder(
                account1,
                account2
        );

        assertThat(accountRepository.findByClientId(client2.getId())).isEmpty();
    }

}

File: src/test/kotlin/io/pivotal/kotlindemo/kotlin/accounts/AccountRepositoryTest.kt
package io.pivotal.kotlindemo.kotlin.accounts

import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test

abstract class AccountRepositoryTest {
    abstract fun getAccountRepository(): AccountRepository

    private lateinit var accountRepository: AccountRepository

    @Before
    fun setup() {
        accountRepository = getAccountRepository()
    }

    @Test
    fun `create() adds accounts with unique ids`() {
        val createAccountRequest1 = buildCreateAccountRequest(
            clientId = 1L,
            number = "account 1 number",
            fullname = "account 1 full name",
            nickname = "account 1 nickname",
            type = AccountType.CREDIT
        )

        val account1 = accountRepository.create(createAccountRequest1)

        assertThat(account1).isEqualToIgnoringGivenFields(
            buildAccount(
                clientId = createAccountRequest1.clientId,
                number = createAccountRequest1.number,
                fullname = createAccountRequest1.fullname,
                nickname = createAccountRequest1.nickname,
                type = createAccountRequest1.type
            ),
            "id"
        )

        val createAccountRequest2 = buildCreateAccountRequest(
            clientId = 2L,
            number = "account 2 number",
            fullname = "account 2 full name",
            nickname = "account 2 nickname",
            type = AccountType.INVESTMENT
        )

        val account2 = accountRepository.create(createAccountRequest2)

        assertThat(account2).isEqualToIgnoringGivenFields(
            buildAccount(
                clientId = createAccountRequest2.clientId,
                number = createAccountRequest2.number,
                fullname = createAccountRequest2.fullname,
                nickname = createAccountRequest2.nickname,
                type = createAccountRequest2.type
            ),
            "id"
        )

        assertThat(account1.id).isNotEqualTo(account2.id)
    }

    @Test
    fun `findAll() returns all created accounts`() {
        val account1 = accountRepository.create(buildCreateAccountRequest())
        val account2 = accountRepository.create(buildCreateAccountRequest())

        assertThat(accountRepository.findAll()).containsExactlyInAnyOrder(account1, account2)
    }

    @Test
    fun `findById(), when the account exists, returns the account with the given id`() {
        val account1 = accountRepository.create(buildCreateAccountRequest())
        val account2 = accountRepository.create(buildCreateAccountRequest())

        assertThat(accountRepository.findById(account1.id)).isEqualTo(account1)
        assertThat(accountRepository.findById(account2.id)).isEqualTo(account2)
    }

    @Test(expected = NoAccountFoundException::class)
    fun `findById(), when the account does not exist, throws a NoAccountFoundException`() {
        val nonExistantAccountId = 12345L
        accountRepository.findById(id = nonExistantAccountId)
    }

    @Test
    fun `findByClientId() returns a (possibly empty) list of all accounts for the given client id`() {
        val account1 = accountRepository.create(buildCreateAccountRequest(clientId = 1L))
        val account2 = accountRepository.create(buildCreateAccountRequest(clientId = 1L))

        assertThat(accountRepository.findByClientId(1L)).containsExactlyInAnyOrder(account1, account2)
        assertThat(accountRepository.findByClientId(2L)).isEmpty()
    }
}
